mempool/src/tests/integration_tests.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::tests::test_framework::{
+    respond_to_request, send_message, test_transactions, MempoolTestFramework,
+};
+use diem_config::{config::PeerRole, network_id::NetworkId};
+use diem_types::PeerId;
+use futures::executor::block_on;
+use netcore::transport::ConnectionOrigin;
+use network::{
+    protocols::wire::handshake::v1::ProtocolIdSet,
+    testutils::{
+        builder::TestFrameworkBuilder,
+        test_framework::TestFramework,
+        test_node::{send_next_network_msg, NodeId, TestNode},
+    },
+    transport::ConnectionMetadata,
+    ProtocolId,
+};
+
+#[test]
+fn single_node_test() {
+    let mut test_framework: MempoolTestFramework = TestFrameworkBuilder::new()
+        .add_owners(1)
+        .add_validator(0)
+        .build();
+    let mut node = test_framework.take_node(NodeId::validator(0));
+    let network_id = NetworkId::Validator;
+    let other_peer_id = PeerId::random();
+    let mut metadata = ConnectionMetadata::mock_with_role_and_origin(
+        other_peer_id,
+        PeerRole::Validator,
+        ConnectionOrigin::Outbound,
+    );
+    metadata.application_protocols = ProtocolIdSet::all_known();
+    let future = async move {
+        let test_txns = test_transactions(0, 1);
+        let inbound_handle = node.node_inbound_handle(network_id);
+        node.assert_txns_not_in_mempool(&test_transactions(0, 3));
+        node.add_txns_via_client(&test_txns).await;
+
+        // After we connect, we should try to send messages to it
+        inbound_handle.connect(
+            node.peer_network_id(network_id).peer_id(),
+            network_id,
+            metadata,
+        );
+
+        // Respond and at this point, all txns should be good to go
+        respond_to_request(&mut node, network_id, other_peer_id, &test_txns).await;
+        let test_txns = test_transactions(1, 1);
+        node.add_txns_via_client(&test_txns).await;
+        node.assert_txns_in_mempool(&test_transactions(0, 2));
+        respond_to_request(&mut node, network_id, other_peer_id, &test_txns).await;
+
+        // Let's also send it an incoming request with more txns and respond with an ack (DirectSend)
+        send_message(
+            &mut node,
+            ProtocolId::MempoolDirectSend,
+            network_id,
+            other_peer_id,
+            &test_transactions(2, 1),
+        )
+        .await;
+        node.assert_txns_in_mempool(&test_transactions(0, 3));
+        node.commit_txns(&test_transactions(0, 3));
+        node.assert_txns_not_in_mempool(&test_transactions(0, 3));
+    };
+    block_on(future);
```

Comment:
You can avoid this `block_on` by using `#[tokio::test]`. The entire test can be async and you can use await inside it.

Reply:
Yeah, let me see if I can do that now because I rewrote the entire test framework.

The original framework was launching runtimes, which then would launch inside a runtime (which then got the tokio runtime very very angry).

Reply:
Yeah, it's still blocking on an error from that, so as much as it would make it a little cleaner, it won't happen unless I refactor it to replace the runtime with the tokio runtime right now.
```
Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context.
thread 'tests::integration_tests::fn_to_val_test' panicked at 'Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context.', /opt/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.9.0/src/runtime/blocking/shutdown.rs:51:21
```

Reply:
Aah, annoying. Okay, sounds good. For now, we can leave it as it is 😄

Reply:
I'll fix this in the next PR.